### PR TITLE
Chapter2

### PR DESCRIPTION
**Pull request by asjad02**
### Solutions for Chapter2

Thanks for helping in refining my solution for the previous chapter and especially for this initiative

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter2.hs [hunk 1]
```diff
@@ -605,8 +626,8 @@ Implement a function that duplicates each element of the list
 
 -}
 duplicate :: [a] -> [a]
-duplicate = error "duplicate: Not implemented!"
-
+duplicate [] =  []
+duplicate (x:xs) = x : x : duplicate(xs)
```
[the remark this answers is not in the document]
Comment (chshersh) [reply src/Chapter2.hs:630]: Hintman, you are drunk, go home :beers:

File: src/Chapter2.hs [hunk 2]
```diff
@@ -620,7 +641,10 @@ Write a function that takes elements of a list only on even positions.
 >>> takeEven [2, 1, 3, 5, 4]
 [2,3,4]
 -}
-takeEven = error "takeEven: Not implemented!"
+takeEven :: [a] -> [a]
+takeEven [] = []
+takeEven [x] = [x]
+takeEven (x: _: xs) = x : takeEven xs
```
Comment (chshersh) [inline src/Chapter2.hs:647]: Very nice solution! Pattern-matching award :trophy:

File: src/Chapter2.hs [hunk 3]
```diff
@@ -857,7 +886,9 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
+rewind :: [a] -> [a]
+rewind [] = []
+rewind (x: xs) = rewind xs ++ [x]
```
Comment (chshersh) [inline src/Chapter2.hs:891]: Your solution is correct! However, it is slow. In lists, it is quite slow to add anything at the end of the list. That is why it is always better to rewrite it with the `:` cons. Remember the explanation with trains? 🚂 🚋 🚋 

That is why a more efficient solution is with the accumulator and the recursive function that will do the addition at the start of the list which is instant!

You can read a bit more about the `go` pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#recursive-go

**Review by chshersh — APPROVED** [review]
Nice work! And well done on solving the Advanced task :clap: 

I left some comments. And Hintman suggested some changes as well, you may find helpful :slightly_smiling_face: 

I'm glad you find our course useful and interesting :blush: Thanks for kind words about it!

**Comment by chshersh** [comment]
@asjad02 Let us know if you want to work on the Advanced Task in this PR :slightly_smiling_face: Otherwise, you can go ahead and close it :checkered_flag: It still counts towards your Hacktoberfest contribution, so don't worry! But, of course, we can't merge it :no_entry:

**Comment by asjad02** [comment]
Hi @chshersh 

Yesterday it got too late and thus couldn't attempt the Final Task.But, will try this weekend :).
Actually, instead of commit I went ahead with commit and push -> sorry for the hassles.

Furthermore, I later realized the PR isn't open in the forked one for chapter 2 and unfortunately cant change base branch with the forked one.(will raised forked master's PR with final solution soon and close this one and attach this PR's link -> just for my own reference)

Thus, will try to attempt to cover all these things in this timespan. 
Thanks for the suggested material.

**Pull request closed without merging**